### PR TITLE
Use Path::real_parent instead of Path::parent

### PR DESCRIPTION
**Pull request by tesujimath**
# Description
For discussion and motivation see #12978 and the [rationale](https://docs.rs/real_parent/latest/real_parent/#rationale) in the real_parent crate docs.

I will follow up with a PR to change `canonicalize` into `make_absolute`, which requires `real_parent` instead of `parent` to be safe in the presence of symlinks.

Note that real_parent touches the filesystem and requires the path to exist, so certain uses of `Path::parent` must remain.
Also, unable to use it in nu-test-support without breaking the test `commands::ucp::copies_the_file_inside_directory_if_path_to_copy_is_directory` which switches cwd out from under its feet (I think).

# User-Facing Changes
Should be no noticeable changes at all, at least until the follow-up PR.

# Tests + Formatting
All these tests passed:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use toolkit.nu; toolkit test stdlib"` to run the tests for the standard library

### CODE REVIEW

**Comment by tesujimath** [comment]
My follow-up draft PR (which wanted to be a stacked PR in this repo) is tesujimath/nushell#1

I have a few remaining tests to fix up, and will be test-driving these changes, but making it visible early.

File: crates/nu-cli/src/config_files.rs [hunk 1]
```diff
@@ -180,9 +182,9 @@ pub fn add_plugin_file(
     if let Ok(cwd) = engine_state.cwd_as_string(None) {
         if let Some(plugin_file) = plugin_file {
             let path = Path::new(&plugin_file.item);
-            let path_dir = path.parent().unwrap_or(path);
+            let path_dir = path.real_parent().unwrap_or(path.to_path_buf());
```
Comment (devyn) [inline crates/nu-cli/src/config_files.rs:185]: should be `unwrap_or_else()`

**Comment by devyn** [comment]
I'm not sure I completely understand how this works. It would be helpful if you could provide an example of what behaviour would be changed by this.

**Comment by tesujimath** [comment]
> I'm not sure I completely understand how this works. It would be helpful if you could provide an example of what behaviour would be changed by this.

This is an enabler to safely handle non-canonicalized paths. So, nothing changes until the follow-up PR, which I have linked. 

The underlying problem being solved is that `Path::parent` is broken in the presence of symlinks. There are two solutions. 

1. As present, eagerly canonicalize all paths, so no symlinks remain.

2. Replace calls to `parent` with `real_parent` and keep paths as they are without canonicalization.

I have presented rationale in my crate docs why I believe that 2 is more ergonomic, which is why I would like to see this in Nushell.

**Comment by kubouch** [comment]
It would be great if you could provide some failing cases in Nushell which would be solved with your changes. Right now, all the rationale makes sense, but it's not clear how it affects Nushell: More specifically, which concrete failing cases of Nushell would your `real_path` solve?

I came up with some:
```
mkdir symlink-test
cd symlink-test

mkdir foo
mkdir bar/baz
ln -s foo bar/baz/foo-link
cd bar/baz/foo-link

ls ..
# lists "foo" and "bar", the contents of "symlink-test"

'..' | path expand --no-symlink
# prints path to "symlink-test"
```

One more note, if we decide to go forward with it, the `real_path` crate must make sure to not touch `std::env` at all because we use our own environment (can happen e.g. with `path.canonicalize()` or when resolving `.`). The only case when it's permitted is when the path is guaranteed to be absolute. That's one reason we wrote our own `nu-path` crate.

**Comment by tesujimath** [comment]
> It would be great if you could provide some failing cases in Nushell which would be solved with your changes. 

So currently there are no failing cases in Nushell, because they have been mitigated by eagerly calling `Path::canonicalize`.
This PR by itself is merely an enabler for the [one I'm really interested in](https://github.com/tesujimath/nushell/pull/1), which removes this eager path canonicalization. I'm keeping that as a kind of stacked PR in my own repo, for separation of concerns to ease the review process.  And the follow-up PR has some failing tests that I need to resolve, although I am happily daily driving it right now.

Briefly, eager canonicalization is not ergonomic, because it exposes symlinks which ought not to be exposed (kind of violation of abstraction).  This would surface wherever a path is stored (e.g. plugins registry), where I suggest should be absolute but not canonical, e.g. `/home/sjg/.config/nushell/plugins/nu_plugin_bash_env` rather than `/nix/store/2z0lixsf8accyf19pdh36lr9jl7sr9fd-home-manager-files/.config/nushell/plugins/nu_plugin_bash_env`, because what business does Nushell have to delve into my symlinks?  

> One more note, if we decide to go forward with it, the `real_path` crate must make sure to not touch `std::env` at all because we use our own environment (can happen e.g. with `path.canonicalize()` or when resolving `.`). The only case when it's permitted is when the path is guaranteed to be absolute. That's one reason we wrote our own `nu-path` crate.

So I'm not really understanding this point.  `real_parent` does indeed open files by relative path, which relies on the process-wide current working directory.  It's not that it's getting this from a `std::env` function, it's fundamental behaviour of how filesystems work, that relative paths are a thing.

Is there some prior Nushell discussion on this point that you could point me at?

Thanks for engaging with me on this.  Very much hoping we can come to a common understanding!

**Comment by kubouch** [comment]
So you're essentially proposing replacing _all_ canonical paths by absolute paths, right? (Unless asked specifically, like `path expand`.) I think that would clean up a lot of the path handling in Nushell, I'm on board with that.

What your PRs and explanations lack is a clear description of what would change for Nushell users. E.g., I already made two examples above, I guess these would be fixed? In the draft PR you linked, I'm missing some “Before: ..., After: ...”. 

I'm not yet sold on using the `real_parent` crate, however, for the following reasons:

1.
> So I'm not really understanding this point. **real_parent does indeed open files by relative path, which relies on the process-wide current working directory**. It's not that it's getting this from a std::env function, it's fundamental behaviour of how filesystems POSIX process environments work, that relative paths are a thing.

This is problematic in Nushell because our `cd` doesn't change Nushell's process working directory. It only changes the PWD environment variable and our environment variables are scoped and thread-safe, see also https://www.nushell.sh/book/thinking_in_nu.html#nushell-s-environment-is-scoped. Any path expansion we do must be done on an absolute path; otherwise the expansion will use the process current working directory and produce an incorrect result. That's why we use `canonicalize_with()` everywhere, but no `canonicalize()` -- you always need to supply the current directory manually read from `$env.PWD` (which is guaranteed to be absolute). We'd need to make sure that all calls to `.real_parent()` happen on an absolute path, or change it to take a cwd parameter.

2.
Why can't `real_parent()` operate lexically? The cases listed in https://docs.rs/real_parent/latest/real_parent/trait.PathExt.html#tymethod.real_parent seem to me like they could be solved lexically as well (e.g., parent of `..` being `../..`). I don't understand the statement “`Path::parent` is not safe in the presence of symlinks”.

For example, take `ls ..`, assuming `$env.PWD = /foo/bar`. Then, what we'd do is join these paths into `/foo/bar/..`, then take a lexical parent, ending up with `/foo/bar`. No need to touch the file system or know anything about the symlinks. I guess the reason my earlier example failed was that we canonicalize the path somewhere during the process. Can't we just use [`std::path::absolute()`](https://doc.rust-lang.org/std/path/fn.absolute.html) instead of `canonicalize()`?

3.
There is also a performance penalty of touching the filesystem vs. operating only lexically. Not sure how much impact it has on common user code, but could be measurable if you do a lot of path manipulations.

**Comment by tesujimath** [comment]
Yes, proposing essentially no longer using canonicalize anywhere at all. 

There's quite a lot to say here, so please take this as a timely but incomplete answer. 

`real_parent` absolutely cannot be a purely lexical thing. The [Getting Dot-Dot Right](https://9p.io/sys/doc/lexnames.html) paper explains well. Briefly, if the last component of the path is a symlink, the real parent is the parent of the symlink target. That can only be done by reading the filesystem and following symlinks.

This really concerns me though: 

> This is problematic in Nushell because our cd doesn't change Nushell's process working directory.

That looks like a serious design error to me. I think many other difficulties stem from that decision. So my initial response, with apologies for not knowing the history of that decision, is to suggest that needs fixing, so that Nu's `cd` actually changes the process cwd. We may need to discuss that further!! 😁

Sorry, little time just now.

**Comment by devyn** [comment]
> That looks like a serious design error to me. I think many other difficulties stem from that decision. So my initial response, with apologies for not knowing the history of that decision, is to suggest that needs fixing, so that Nu's cd actually changes the process cwd. We may need to discuss that further!! 😁

Yes, it's actually very simple: in a traditional POSIX shell, subshells are actually done by forking, and nothing is ever done in parallel within the process, so changing process working directory is fine. But that isn't the case for Nushell - we operate more like a traditional scripting language, and we handle scoped environment and variables ourselves.

This works perfectly fine in Nushell, without spawning any subprocesses:

```nushell
glob --no-file * | par-each { |dir|
  cd $dir
  cp Cargo.toml Cargo.toml.bak
  open Cargo.toml.bak | update dependencies.mimalloc.version "0.1.43" | save Cargo.toml
}
```

Of course, this is a totally contrived example and the `cd` isn't totally necessary, but we think it's less of a gotcha for closures to have completely isolated scope. We handle this by setting `$env.PWD` within the scope, which all commands are supposed to read and construct paths as relative to when doing filesystem operations.

We generally don't change the environment variables or working directory of the `nu` process, so those changes are only visible when spawning external processes, at which point those are collected from the state of the scope that spawned them.

**Comment by tesujimath** [comment]
Wow, that's a big surprise to me!

It means you have to give up on relative paths altogether, which is a native feature provided by the kernel. So you're having to implement operating system level functionality inside Nushell itself. 

I think this is a terrible mistake. The benefit of pseudo-thread-local cwd that you simulate by working really hard to make it look like you have real separation can't be with worth the price in complexity being paid. Can it?

**Comment by tesujimath** [comment]
Just want to clarify something here, in case my previous comment sounded a bit disrespectful, which certainly isn't my intention.

I think the Nushell model of cwd doesn't match what many people would expect from a shell.  Certainly I was surprised not to find the underlying OS functionality of process-wide cwd followed in Nushell. But if what I'm saying here isn't striking a chord, I will gently withdraw my PR and leave things with you.  My intention has been to represent a view for consideration.

Without a process-wide cwd, the `real_parent` crate won't be useful, and what I have been trying to achieve here may be unrealistic.

**Comment by kubouch** [comment]
We used to change the process working directory with `cd`, but we removed it because modifying a global state from inner scopes doesn't fit Nushell's design, and it was causing bugs. To support changing the process cwd, we would need to:
* Give up parallel processing (`par-each`)
* Break our design principle of no mutation of outer scope...
* ...or, to keep the scoped cwd, we'd need to maintain some kind of stack of entered cwds and keep changing them when entering and leaving blocks or closures. This would add complexity.

To sidestep the above issues, for us, doing manual expansion with `$env.PWD` is worth the extra effort.

> `real_parent` absolutely cannot be a purely lexical thing. The [Getting Dot-Dot Right](https://9p.io/sys/doc/lexnames.html) paper explains well. Briefly, if the last component of the path is a symlink, the real parent is the parent of the symlink target. That can only be done by reading the filesystem and following symlinks.

Could you come up with an example where resolving a parent cannot be done lexically? I can't think of any. The motivation examples in the paper should be solvable by using absolute, non-canonical paths. I don't have time to read it carefully at the moment.

E.g., consider an absolute path `cd /foo/bar/baz-link` where `baz-link` is a symlink to `/a/b/baz`. Then, `cd ..` would be lexically resolved as `/foo/bar`. That's correct, right? If we operate on absolute, not canonicalized paths, then it wouldn't make sense to me for `cd ..` to magically jump to `/a/b`.

Even if this PR won't land, what we can take from your efforts is that we could use absolute paths everywhere instead of canonical. And I think the `real_parent` will still be useful for others, even if we don't find a use for it in Nushell.

**Comment by kubouch** [comment]
Thanks, that's a good example. It's directly applicable to Nushell modules, for example,
```nushell
# spam.nu
use foo.nu
def main [] { foo }
```
However, I think in this case, we'd still want the `foo.nu` to be looked up relative to the symlink, not the symlink target. To make it work, you'd need to symlink both `spam.nu` and `foo.nu`, or symlink the whole directory. If we switch to “all paths are absolute without resolved symlinks”, considering imports relative to the symlink target would break the consistency. It would also mean that you could symlink, e.g., only one file from a multi-file module, and it would keep working, that would be strange. Note that if you really want to use the canonical path, you could use `path expand` to resolve the symlinks, but going the opposite direction -- from symlink target to the symlink -- would be much more difficult.

I'll close this PR, then, but thanks for the input, it prompted us to rethink the absolute vs. canonical paths.

**Pull request closed without merging**